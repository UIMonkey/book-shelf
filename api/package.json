{
  "name": "api",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "ts-node-dev src/index.ts",
    "test": "echo \"Error: no test specified\" && exit 1",
    "build": "tsc"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {},
  "devDependencies": {
    "ts-node-dev": "^1.1.1",
    "typescript": "^4.1.3"
  }
}
