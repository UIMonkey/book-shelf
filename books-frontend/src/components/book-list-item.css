.book-list-item {
    background-color: #62efff;
    border-radius: 5px;
    box-shadow: rgba(0, 0, 0, 0.16) 0px 1px 4px;
    cursor: pointer;
    margin: 0px auto 4px auto;
    padding: 8px;
    width: 325px;
    display: flex;
}

.item-param {
    display: block;
}
