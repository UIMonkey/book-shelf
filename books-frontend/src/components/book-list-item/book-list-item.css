.book-list-item {
  border-radius: 5px;
  box-shadow: rgba(0, 0, 0, 0.16) 0px 1px 4px;
  cursor: pointer;
  margin: 0px auto 6px auto;
  padding: 8px;
  width: 384px;
  display: flex;
  transform: scale(1);
  transition: transform 0.1s ease-in-out;
}

.book-list-item__img {
  margin: 8px 16px 8px 16px;
}

.book-list-item:hover {
    box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;
  transform: scale(1.02);
}

.item-param {
  display: block;
}
