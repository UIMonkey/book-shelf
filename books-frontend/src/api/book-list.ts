import { ApolloError } from "@apollo/client";
import { IVolume } from "../../../api/build";

export interface IBookList {
    loading: boolean;
    error: ApolloError | undefined;
    data: any;
    handleSelect: (book: IVolume) => void
}

export interface IBookListItem {
    book: IVolume;
    selectBook: (book: IVolume) => void
}
