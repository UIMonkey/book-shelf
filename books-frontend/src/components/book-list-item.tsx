import { IBookListItem } from "../api/book-list";
import { titleCase } from "../utils/utils";
import './book-list-item.css';

export const BookListItem = (props: IBookListItem) => (
    <div className="book-list-item" onClick={(_) => props.selectBook(props.book)}>
        <img src={props?.book?.volumeInfo?.imageLinks?.smallThumbnail} alt="Book"></img>
        <div>
            <h3 className="item-param" data-testid="title">{titleCase(props.book?.volumeInfo?.title)}</h3>
            {/* <span className="item-param">{titleCase(props.book.volumeInfo?.authors)}</span> */}
            <span className="item-param" data-testid="subtitle">{titleCase(props.book?.volumeInfo?.subtitle)}</span>
        </div>
    </div>
)

export default BookListItem;