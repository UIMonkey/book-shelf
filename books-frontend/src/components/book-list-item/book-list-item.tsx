import { IBookListItem } from "../../api/book-list";
import { titleCase } from "../../utils/utils";
import './book-list-item.css';

export const BookListItem = (props: IBookListItem) => (
    <div className="book-list-item" onClick={(_) => props.selectBook(props.book)}>
        <img
            src={props?.book?.volumeInfo?.imageLinks?.smallThumbnail}
            alt="Book"
            width="64"
            className="img-rounded book-list-item__img"></img>
        <div>
            <h4 className="item-param" data-testid="title">{titleCase(props.book?.volumeInfo?.title)}</h4>
            <p className="item-param lead" data-testid="subtitle">{titleCase(props.book?.volumeInfo?.subtitle)}</p>
        </div>
    </div>
)

export default BookListItem;