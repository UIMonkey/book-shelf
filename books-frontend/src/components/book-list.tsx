import BookListItem from './book-list-item/book-list-item';
import { IBookList } from "../api/book-list";
import { IVolume } from "../../../api/build";
import React from 'react';
import LoadingSpinner from './loading-spinner/loading-spinner';

export const BookList = (props: IBookList) => {

    if (props.loading) {
        return (
            <LoadingSpinner />
        )
    }

    if (props.error) {
        return (
            <h1>Failed to Fetch</h1>
        )
    }

    return (
        <div className="col-md-6" style={{ height: '100%', overflow: "auto" }}>
            {props?.data?.books?.map((book: IVolume) => (
                <BookListItem key={book?.volumeInfo?.title} book={book} selectBook={props.handleSelect}></BookListItem>
            ))}
        </div>
    )
}


export default BookList;