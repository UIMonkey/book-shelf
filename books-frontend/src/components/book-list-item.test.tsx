import { render, screen } from '@testing-library/react';
import { IVolume, Volume } from '../../../api/build';
import { IBookListItem } from '../api/book-list';
import { BookListItem } from './book-list-item';

const testVolume = new Volume('hjskdf', 'This Is My Book');

const testBook: IBookListItem = {
    book: {
        ...testVolume,
        volumeInfo: {
            ...testVolume.volumeInfo,
            subtitle: 'I am subtitle',
            authors: ['John', 'Sue']
        }
    },
    selectBook: () => { }
}

describe('BookListItem', () => {

    test('It renders the booklistitem', () => {
        render((<BookListItem key={testBook.book.id} book={testBook.book} selectBook={testBook.selectBook} />));
    });

    test('It renders the book title', () => {
        render(<BookListItem key={testBook.book.volumeInfo.title} book={testBook.book} selectBook={testBook.selectBook}></BookListItem>);
        expect(screen.getByText(testVolume.volumeInfo.title)).toBeInTheDocument();
    });

    test('The title should always be in titlecase', () => {
        const { getByTestId } = render(<BookListItem key={testBook.book.volumeInfo.title} book={testBook.book} selectBook={testBook.selectBook}></BookListItem>);
        const title = getByTestId('title');
        const wordArray = title.textContent?.split(' ');
        const firstLetterRegEx = /^[A-Z]/;
        if (wordArray) {
            expect(wordArray.every((word) => firstLetterRegEx.test(word))).toBe(true);
        } else {
            expect(true).toBe(true);
        }
    });

    test('It renders the subtitle', () => {
        const { getByTestId } = render(<BookListItem key={testBook.book.volumeInfo.title} book={testBook.book} selectBook={testBook.selectBook}></BookListItem>);
        const subtitle = getByTestId('subtitle');
        expect(subtitle).toHaveTextContent('I Am Subtitle');
    });

    test('It matches the snapshot', () => {
        const { container, getByText } = render(<BookListItem key={testBook.book.volumeInfo.title} book={testBook.book} selectBook={testBook.selectBook}></BookListItem>);
        expect(getByText(testVolume.volumeInfo.title)).toBeInTheDocument();
        expect(container.firstChild).toMatchSnapshot();
    });
});