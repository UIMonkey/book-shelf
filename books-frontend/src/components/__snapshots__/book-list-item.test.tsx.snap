// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListItem It matches the snapshot 1`] = `
<div
  class="book-list-item"
>
  <img
    alt="Book image"
    src=""
  />
  <div>
    <h3
      class="item-param"
      data-testid="title"
    >
      This Is My Book
    </h3>
    <span
      class="item-param"
      data-testid="subtitle"
    >
      I Am Subtitle
    </span>
  </div>
</div>
`;
